Fix Atreian Passport system and updated for retail

SVDNESS opened the pull request:
Atreian Passport System turned out to be incomplete, and there were several critical bugs related to receiving rewards (double or infinite), as well as problems with synchronizing the time record in the database.
There was no implementation of attributes such as "reward_permit_level" and "reward_item_expire_time", as well as no implementation for the "ANNIVERSARY" type.
Fixed all the errors and implemented the system to the end, as it should work for retail.

SVDNESS committed:
Fix Atreian Passport system and updated for retail

neon-dev committed:
Autoformat & move comments to SM_SYSTEM_MESSAGE

Also reverted AuditLogger changes, because the messages resolve to "<Player> tried to ..." after formatting.

neon-dev commented on game-server/src/com/aionemu/gameserver/network/aion/serverpackets/SM_SYSTEM_MESSAGE.java:
```diff
@@ -27859,6 +27859,13 @@ public static SM_SYSTEM_MESSAGE STR_REGISTER_ITEM_MSG_UPGRADE_CANNOT_NEED_QINA()
+	 * You can only obtain %1% when you have reached Level %0.
+	 */
+	public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(String name, int level) {
+		return new SM_SYSTEM_MESSAGE(1402573, name, level);
```
`name` and `level` must be switched to display correctly:
```suggestion
	public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(int level, String name) {
		return new SM_SYSTEM_MESSAGE(1402573, level, name);
```

neon-dev replied:
> In client:
>```xml
><string>
>  <id>1402573</id>
>  <name>STR_MSG_ATTEND_REWARD_INVALID_LEVEL</name>
>  <body>You can only obtain %1% when you have reached Level %0.</body>
></string>
>```
>The NAME attribute is passed first, then the LEVEL.
>Therefore, it is true. STRING then INT.
>```java
>public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(String name, int level) {
>    return new SM_SYSTEM_MESSAGE(1402573, name, level);
>}
>```
>And the challenge:
>PacketSendUtility.sendPacket(player,
>SM_SYSTEM_MESSAGE.STR_MSG_ATTEND_REWARD_INVALID_LEVEL(itemName, minLevel));
>And it works correctly in the game.

%0 references the first parameter and %1% references the second parameter. If name is passed first, it will be displayed as the level, because it uses %0. This is true for all official client languages. If your client language renders it differently, it would be due to a modded (custom) language pack.

neon-dev commented on game-server/src/com/aionemu/gameserver/model/account/PassportsList.java:
```diff
@@ -40,4 +42,17 @@ public boolean isPassportPresent(int passportId) {
+			if (pp.getId() != passportId) {
+				continue;
+			}
+			var ppDay = pp.getArriveDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
```
`ZoneId.systemDefault()` can lead to mismatching dates if the system timezone is different from the configured game server timezone.
This should always work correctly:
```suggestion
			var ppDay = ServerTime.atDate(pp.getArriveDate()).toLocalDate();
```

SVDNESS replied:
Yes, you're right, you should set the server time.
The final method will be this 1 or 2 option, depending on the writing style:
    public boolean hasPassportForDay(int passportId, LocalDate attendDay) {
        for (var pp : passports) {
            if (pp.getId() != passportId) {
                continue;
            }
            var ppDay = ServerTime.atDate(pp.getArriveDate()).toLocalDate();
            if (ppDay.equals(attendDay)) {
                return true;
            }
        }
        return false;
    }
Or:
    public boolean hasPassportForDay(int passportId, LocalDate attendDay) {
        return passports.stream()
                .filter(pp -> pp.getId() == passportId)
                .map(pp -> ServerTime.atDate(pp.getArriveDate()).toLocalDate())
                .anyMatch(attendDay::equals);
    }

neon-dev commented on game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java:
```diff
@@ -71,143 +76,237 @@ private boolean isAtreianPassportDisabled(LocalDateTime checkDateTime) {
+			return true;
+		}
+		var lastAttendDay = ServerTime.atDate(last).toLocalDateTime().minusHours(ATTEND_RESET_HOUR).toLocalDate();
+		var currentAttendDay = now.minusHours(ATTEND_RESET_HOUR).toLocalDate();
```
You have created the helper function `getAttendDay` below, which should be used here:
```suggestion
		var lastAttendDay = getAttendDay(ServerTime.atDate(last).toLocalDateTime());
		var currentAttendDay = getAttendDay(now);
```

SVDNESS replied:
Yes, you're right, it's the right thing to do:
    private boolean checkOnlineDate(Account pa, LocalDateTime now) {
        var last = pa.getLastStamp();
        if (last == null) {
            return true;
        }
        var lastAttendDay = getAttendDay(ServerTime.atDate(last).toLocalDateTime());
        var currentAttendDay = getAttendDay(now);
        return !currentAttendDay.equals(lastAttendDay);
    }

neon-dev commented on game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java:
```diff
@@ -71,143 +76,237 @@ private boolean isAtreianPassportDisabled(LocalDateTime checkDateTime) {
-	/*
-	 * Maximum 50 Passports are saved
-	 */
+	//Сохраняется не более 50 паспортов.
```
I guess this can be removed? If not, please translate it and apply the JavaDoc format (or move the single line comment inside the method).

SVDNESS replied:
private void checkPassportLimit(Player player) {
        var pa = player.getAccount();
        var pl = pa.getPassportsList().getAllPassports();
        //More than 50 passports cannot be saved.
        if (pl.size() < 50) {
            return;
        }
        var oldest = pl.stream()
                .filter(pp -> !pp.isFakeStamp())
                .min(Comparator.comparing(Passport::getArriveDate))
                .orElse(null);
        if (oldest == null) {
            oldest = pl.stream()
                    .min(Comparator.comparing(Passport::getArriveDate))
                    .orElse(null);
        }
        if (oldest != null) {
            oldest.setPersistentState(PersistentState.DELETED);
            pa.getPassportsList().removePassport(oldest);
            AccountPassportsDAO.storePassportList(pa.getId(), List.of(oldest));
            PacketSendUtility.sendPacket(player, SM_SYSTEM_MESSAGE.STR_MSG_ATTEND_REWARD_REMOVE_EXCESS("Atreian Passport"));
        }
    }

neon-dev commented on game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java:
```diff
@@ -71,143 +76,237 @@ private boolean isAtreianPassportDisabled(LocalDateTime checkDateTime) {
+
+	private int getAccountAgeInMonths(LocalDate creationDate, LocalDate now) {
+		int months = (now.getYear() - creationDate.getYear()) * 12 + (now.getMonthValue() - creationDate.getMonthValue());
+		//Если сегодня день месяца меньше дня создания — полный месяц еще не прошел.
```
This should be translated as well.

SVDNESS replied:
/**
     * Calculates the number of full months between the account creation date and the given date.
     * <p>
     * The calculation is based on year and month difference. If the day of the given date is
     * earlier than the day of the creation date, the current month is considered incomplete
     * and is not counted.
     * <p>
     * The returned value is always non-negative.
     *
     * @param creationDate the date when the account was created
     * @param now the current date used for the calculation
     * @return the number of full months since account creation
     */
     private int getAccountAgeInMonths(Player player, LocalDate now) {...}

neon-dev commented on game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java:
```diff
@@ -71,143 +76,237 @@ private boolean isAtreianPassportDisabled(LocalDateTime checkDateTime) {
+		return ServerTime.atDate(player.getCreationDate()).toLocalDate();
+	}
+
+	private int getAccountAgeInMonths(LocalDate creationDate, LocalDate now) {
```
The method is called `getAccountAgeInMonths` but is not explicitly handling account data. It could be changed to accept the player object instead of the creationDate and process the date conversion from line 187 in here. Then the method name would be accurate.

SVDNESS replied:
New method:
 private int getAccountAgeInMonths(Player player, LocalDate now) {
        var creationDate = ServerTime.atDate(player.getCreationDate()).toLocalDate();
        int months = (now.getYear() - creationDate.getYear()) * 12 + (now.getMonthValue() - creationDate.getMonthValue());
        if (now.getDayOfMonth() < creationDate.getDayOfMonth()) {
            months--;
        }
        return Math.max(0, months);
    }


And replace (this: case ANNIVERSARY -> {...} :

-var creationDate = getAccountCreationDate(player);
-int monthsAlive = getAccountAgeInMonths(creationDate, now.toLocalDate());
+int monthsAlive = getAccountAgeInMonths(player, now.toLocalDate());
And delete this method:
private LocalDate getAccountCreationDate(Player player) {
return ServerTime.atDate(player.getCreationDate()).toLocalDate();
}

SVDNESS commented:
Hi, I'm responding to your comments.
1.
In client:
<string>
    <id>1402573</id>
    <name>STR_MSG_ATTEND_REWARD_INVALID_LEVEL</name>
    <body>You can only obtain %1% when you have reached Level %0.</body>
  </string>
The NAME attribute is passed first, then the LEVEL.
Therefore, it is true. STRING then INT.
public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(String
name, int level) {
    return new SM_SYSTEM_MESSAGE(1402573, name, level);
}
And the challenge:
PacketSendUtility.sendPacket(player,
SM_SYSTEM_MESSAGE.STR_MSG_ATTEND_REWARD_INVALID_LEVEL(itemName, minLevel));
And it works correctly in the game.

2. var ppDay = ServerTime.atDate(pp.getArriveDate()).toLocalDate(); - yes,
you're right, you should set the server time.
The final method will be this 1 or 2 option, depending on the writing style:
    public boolean hasPassportForDay(int passportId, LocalDate attendDay) {
        for (var pp : passports) {
            if (pp.getId() != passportId) {
                continue;
            }
            var ppDay = ServerTime.atDate(pp.getArriveDate()).toLocalDate();
            if (ppDay.equals(attendDay)) {
                return true;
            }
        }
        return false;
    }
Or:
    public boolean hasPassportForDay(int passportId, LocalDate attendDay) {
        return passports.stream()
                .filter(pp -> pp.getId() == passportId)
                .map(pp ->
ServerTime.atDate(pp.getArriveDate()).toLocalDate())
                .anyMatch(attendDay::equals);
    }

3. Yes, you're right, it's the right thing to do:
    private boolean checkOnlineDate(Account pa, LocalDateTime now) {
        var last = pa.getLastStamp();
        if (last == null) {
            return true;
        }
        var lastAttendDay =
getAttendDay(ServerTime.atDate(last).toLocalDateTime());
        var currentAttendDay = getAttendDay(now);
        return !currentAttendDay.equals(lastAttendDay);
    }

    private LocalDate getAttendDay(LocalDateTime serverTime) {
        return serverTime.minusHours(ATTEND_RESET_HOUR).toLocalDate();
    }


4. This method:
    private void purgeExpiredPassports(Player player) {
        var pa = player.getAccount();
        var ppl = pa.getPassportsList();
        var now = ServerTime.now().toInstant();
        var toRemove = new ArrayList<Passport>();
        for (var pp : new ArrayList<>(ppl.getAllPassports())) {
            if (pp.isRewarded() || pp.isFakeStamp()) {
                continue;
            }
            var atp =
DataManager.ATREIAN_PASSPORT_DATA.getAtreianPassportId(pp.getId());
            if (atp == null) {
                continue;
            }
            int expireMin = atp.getRewardExpireMinutes();
            if (expireMin <= 0) {
                continue;
            }
            var deadline = pp.getArriveDate()
                    .toInstant()
                    .plusSeconds(expireMin * 60L);
            if (now.isAfter(deadline)) {
                pp.setPersistentState(PersistentState.DELETED);
                ppl.removePassport(pp);
                toRemove.add(pp);
            }
        }
        if (!toRemove.isEmpty()) {
            AccountPassportsDAO.storePassportList(pa.getId(), toRemove);
        }
    }
Changing it:
- var now = Instant.now();
+ var now = ServerTime.now().toInstant();
This way there will be a "single time source".

5.
private int getAccountAgeInMonths(LocalDate creationDate, LocalDate now)
{...}

The method calculates the number of full months that have passed since the
account creation date up to the current date.
It first computes the difference in months based on years and months.
Then, if the current day of the month is earlier than the creation day, the
current month is considered incomplete and is not counted.
The result is guaranteed to be non-negative.

   /**
     * Calculates the number of full months between the account creation
date and the given date.
     * <p>
     * The calculation is based on year and month difference. If the day of
the given date is
     * earlier than the day of the creation date, the current month is
considered incomplete
     * and is not counted.
     * <p>
     * The returned value is always non-negative.
     *
     * @param creationDate the date when the account was created
     * @param now the current date used for the calculation
     * @return the number of full months since account creation
     */
     private int getAccountAgeInMonths(Player player, LocalDate now) {
        var creationDate =
ServerTime.atDate(player.getCreationDate()).toLocalDate();
        int months = (now.getYear() - creationDate.getYear()) * 12 +
(now.getMonthValue() - creationDate.getMonthValue());
        if (now.getDayOfMonth() < creationDate.getDayOfMonth()) {
            months--;
        }
        return Math.max(0, months);
    }

And replace:
- var creationDate = getAccountCreationDate(player);
- int monthsAlive = getAccountAgeInMonths(creationDate, now.toLocalDate());
+ int monthsAlive = getAccountAgeInMonths(player, now.toLocalDate());

And delete it:
private LocalDate getAccountCreationDate(Player player) {
    return ServerTime.atDate(player.getCreationDate()).toLocalDate();
}

It seems that these are all corrections based on your comments.

чт, 1 янв. 2026 г. в 08:08, Neon ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In
> game-server/src/com/aionemu/gameserver/network/aion/serverpackets/SM_SYSTEM_MESSAGE.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656084510>
> :
>
> > +	public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(String name, int level) {
> +		return new SM_SYSTEM_MESSAGE(1402573, name, level);
>
> name and level must be switched to display correctly:
> ⬇️ Suggested change
>
> -	public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(String name, int level) {
> -		return new SM_SYSTEM_MESSAGE(1402573, name, level);
> +	public static SM_SYSTEM_MESSAGE STR_MSG_ATTEND_REWARD_INVALID_LEVEL(int level, String name) {
> +		return new SM_SYSTEM_MESSAGE(1402573, level, name);
>
> ------------------------------
>
> In game-server/src/com/aionemu/gameserver/model/account/PassportsList.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656085904>
> :
>
> > @@ -40,4 +42,17 @@ public boolean isPassportPresent(int passportId) {
>  	public List<Passport> getAllPassports() {
>  		return passports;
>  	}
> -}
> +
> +	public boolean hasPassportForDay(int passportId, LocalDate attendDay) {
> +		for (var pp : passports) {
> +			if (pp.getId() != passportId) {
> +				continue;
> +			}
> +			var ppDay = pp.getArriveDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
>
> ZoneId.systemDefault() can lead to mismatching dates if the system
> timezone is different from the configured game server timezone.
> This should always work correctly:
> ⬇️ Suggested change
>
> -			var ppDay = pp.getArriveDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
> +			var ppDay = ServerTime.atDate(pp.getArriveDate()).toLocalDate();
>
> ------------------------------
>
> In
> game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656087341>
> :
>
> > +		var lastAttendDay = ServerTime.atDate(last).toLocalDateTime().minusHours(ATTEND_RESET_HOUR).toLocalDate();
> +		var currentAttendDay = now.minusHours(ATTEND_RESET_HOUR).toLocalDate();
>
> You have created the helper function getAttendDay below, which should be
> used here:
> ⬇️ Suggested change
>
> -		var lastAttendDay = ServerTime.atDate(last).toLocalDateTime().minusHours(ATTEND_RESET_HOUR).toLocalDate();
> -		var currentAttendDay = now.minusHours(ATTEND_RESET_HOUR).toLocalDate();
> +		var lastAttendDay = getAttendDay(ServerTime.atDate(last).toLocalDateTime());
> +		var currentAttendDay = getAttendDay(now);
>
> ------------------------------
>
> In
> game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656088586>
> :
>
> >  	}
>
> -	/*
> -	 * Maximum 50 Passports are saved
> -	 */
> +	//Сохраняется не более 50 паспортов.
>
> I guess this can be removed? If not, please translate it and apply the
> JavaDoc format (or move the single line comment inside the method).
> ------------------------------
>
> In
> game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656089725>
> :
>
> >
> +	private LocalDate getAccountCreationDate(Player player) {
> +		return ServerTime.atDate(player.getCreationDate()).toLocalDate();
> +	}
> +
> +	private int getAccountAgeInMonths(LocalDate creationDate, LocalDate now) {
> +		int months = (now.getYear() - creationDate.getYear()) * 12 + (now.getMonthValue() - creationDate.getMonthValue());
> +		//Если сегодня день месяца меньше дня создания — полный месяц еще не прошел.
>
> This should be translated as well.
> ------------------------------
>
> In
> game-server/src/com/aionemu/gameserver/services/AtreianPassportService.java
> <https://github.com/beyond-aion/aion-server/pull/106#discussion_r2656091185>
> :
>
> >
> +	private LocalDate getAccountCreationDate(Player player) {
> +		return ServerTime.atDate(player.getCreationDate()).toLocalDate();
> +	}
> +
> +	private int getAccountAgeInMonths(LocalDate creationDate, LocalDate now) {
>
> The method is called getAccountAgeInMonths but is not explicitly handling
> account data. It could be changed to accept the player object instead of
> the creationDate and process the date conversion from line 187 in here.
> Then the method name would be accurate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/beyond-aion/aion-server/pull/106#pullrequestreview-3621604791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BH2FW3J5B6KRKSJ3GN7RXTT4ESTTXAVCNFSM6AAAAACQBCS732VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTMMRRGYYDINZZGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

SVDNESS commented:
It doesn't always work that way, I've seen it many times. Probably depends
on the client

> Message ID: ***@***.***
> com>
>

SVDNESS commented:
I've checked it in the game, all the announcements

чт, 1 янв. 2026 г. в 18:54, SVDNESS ***@***.***>:

> It doesn't always work that way, I've seen it many times. Probably depends
> on the client
>
>> Message ID: ***@***.***
>> com>
>>
>

SVDNESS commented:
And I rewrote this method:
private void purgeExpiredPassports(Player player) {
        var pa = player.getAccount();
        var ppl = pa.getPassportsList();
        Instant now = ServerTime.now().toInstant();
        List<Passport> toRemove = new ArrayList<>();
        for (var pp : new ArrayList<>(ppl.getAllPassports())) {
            if (pp.isRewarded() || pp.isFakeStamp()) {
                continue;
            }
            var atp = DataManager.ATREIAN_PASSPORT_DATA.getAtreianPassportId(pp.getId());
            if (atp == null) {
                continue;
            }
            int expireMin = atp.getRewardExpireMinutes();
            if (expireMin <= 0) {
                continue;
            }
            Instant deadline = pp.getArriveDate().toInstant().plusSeconds(expireMin * 60L);
            if (now.isAfter(deadline)) {
                pp.setPersistentState(PersistentState.DELETED);
                ppl.removePassport(pp);
                toRemove.add(pp);
            }
        }
        if (!toRemove.isEmpty()) {
            AccountPassportsDAO.storePassportList(pa.getId(), toRemove);
        }
    }
This way there will be a "single time source"

SVDNESS committed:
Fix Atreian Passport system and updated for retail

neon-dev committed:
Final changes

neon-dev commented on game-server/src/com/aionemu/gameserver/model/account/Passport.java:
```diff
@@ -40,16 +40,17 @@ public void setRewarded(boolean rewarded) {
-		else
-			return rewarded ? RewardStatus.EXPIRED : RewardStatus.AVAILABLE;
+		}
+		return rewarded ? RewardStatus.TAKEN : RewardStatus.AVAILABLE;
```
`RewardStatus.EXPIRED` is no longer used anywhere. I think it should be used when the `reward_item_expire_time` has expired.

neon-dev replied:
Ok, it seems that `RewardStatus.EXPIRED` just hides expired rewards and does not need to be assigned in your implementation, because they are automatically removed via `purgeExpiredPassports`.
Therefore, I think it is finally time to merge your pull request.
